### PR TITLE
feat(common/core/web): input processor unit tests

### PR DESCRIPTION
**Pull request by jahorton**
Since we're going to want some good automated tests for the new predictive text features, I thought it would be wise to get a piece of integrated test support in place that's been missing until now.

Introducing... the first automated test integrating part of Web (`OutputTarget`, `Transcription`) with the LMLayer!  For good measure, there's an extra tidbit that will also be very useful - the test uses a dynamically-compiled model via `@keymanapp/lexical-model-compiler`!  Admittedly, this motivated a few changes through the LMLayer to allow loading from a raw script - I'd rather not leave temp files everywhere.  

_**Motivation for said "few changes"**_: We have some test models that are handwritten and thus must be committed, which would complicated alternative approaches.  Constantly updating .gitignore selectively to cover build products for test resources could easily become a pain, as would reviewing PRs with slight variants on the same model.  Best to commit a backing wordlist just once, rather than once-per-test-variant.

------

As a whole, having all of these changes in place will be very useful for adding proper automated tests to cover #3824.  Note that this branch is completely separate from that one.
- That branch contains a number of features that would result in a number of models with only slight differences... with no change to the actual lexicon / wordlist.

As this PR requires a small change to package.json (devDependency:  @keymanapp/lexical-model-compiler), it's prone to merge conflicts b/c of our daily auto-updates.  So long as that's the only affected file, things should be fine otherwise.

### CODE REVIEW

File: common/core/web/input-processor/test.sh [hunk 1]
```diff
@@ -0,0 +1,71 @@
+#!/bin/bash
+
+# We should work within the script's directory, not the one we were called in.
+cd $(dirname "$BASH_SOURCE")
```
Comment (mcdurdin) [inline common/core/web/input-processor/test.sh:4]: Can we `pushd` and `popd` on exit so that the user's path is preserved?

**Review by mcdurdin — APPROVED** [review]
This seems okay to me.

**Comment by jahorton** [comment]
I think that the Windows (Desktop/Developer) failures are likely due to some sort of `developer/js/package.json` shenanigans during the build process.  I've occasionally seen `npm` tweak package versions during builds unexpectedly, so something might be tweaking package.json's `"version"` entry to something like `"version": "14.0.178-alpha".  This could cause the "version" string to mismatch and thus cause `lerna`'s attempts to link the local development version to fail (due to version string mismatch).

The Developer build process does include steps for `npm pack`ing the lexical-model-compiler (refer to #3166), so if this is somehow causing tweaks to package.json, those need to be reverted before the KMW part of the build commences.

---

Unfortunately, my attempts to address this (and properly investigate it) by building Developer on my local machine are failing.  As a result, builds aren't reaching the point of failure from the CI build logs.  Here's my current point of failure:

```
Keyman.System.BuildLangTags.pas(82): error E2003: Undeclared identifier: 'RaiseExc'
Keyman.Console.BuildStandardsData.pas(13): error F2063: Could not compile used unit 'Keyman.System.BuildLangTags.pas'
```

It has been a while since I've built Developer locally, so maybe there's a new build prerequisite I missed?

In case it matters, UserDefines.mak:
```
DELPHI_VERSION=19.0
GIT_BASH_FOR_KEYMAN="<path>"
SET KEYMAN_ROOT="<path>"
```

**Comment by mcdurdin** [comment]
I took a quick look at the failing build this morning but could not repro on my machine. Will dig deeper as soon as I have time.

**Comment by mcdurdin** [comment]
> It has been a while since I've built Developer locally, so maybe there's a new build prerequisite I missed?

I think you need to update your Delphi version to 10.3.3.

**Comment by mcdurdin** [comment]
> so something might be tweaking package.json's `"version"` entry to something like `"version": "14.0.178-alpha"`. This could cause the "version" string to mismatch and thus cause `lerna`'s attempts to link the local development version to fail (due to version string mismatch).

This is the diff from the agent at the point of the build fail:

```diff
diff --git a/developer/js/package-lock.json b/developer/js/package-lock.json
index 829c59a59..cceb96ed8 100644
--- a/developer/js/package-lock.json
+++ b/developer/js/package-lock.json
@@ -1,6 +1,6 @@
 {
   "name": "@keymanapp/lexical-model-compiler",
-  "version": "14.0.178",
+  "version": "14.0.178-alpha",
   "lockfileVersion": 1,
   "requires": true,
   "dependencies": {
diff --git a/developer/js/package.json b/developer/js/package.json
index e00bafbe2..3a206d467 100644
--- a/developer/js/package.json
+++ b/developer/js/package.json
@@ -1,6 +1,6 @@
 {
   "name": "@keymanapp/lexical-model-compiler",
-  "version": "14.0.178",
+  "version": "14.0.178-alpha",
   "description": "Keyman Developer lexical model compiler",
   "keywords": [
     "keyboard",
```

The version string appears to be updated during the build by developer/js/build.js. I just don't really know why it's falling over now but not in the past.

**Comment by jahorton** [comment]
So it's exactly what I expected.

The reason is this:  in the past, Developer never had to 'build' `@keymanapp/lexical-model-compiler` as part of the KMW build / `npm install`.  This PR changes that, as it's a new `devDependency` of `@keymanapp/input-processor`.

So, before:

1.  The Windows/Developer build is triggered.
2.  That build triggers the lexical model compiler build.
3.  As part of that build, `npm` tweaks the lmc's package.json `version` property.
4.  It is never referenced again within the build.

But now:

1.  The Windows/Developer build is triggered.
2.  That build triggers the lexical model compiler build.
3.  As part of that build, `npm` tweaks the lmc's package.json `version` property.
4.  Some time later, KMW is built (for inclusion within Developer).
5.  KMW triggers a `@keymanapp/input-processor` build.
6.  This build's `package.json` triggers an `install` of `@keymanapp/lexical-model-compiler`... but the changed `version` property causes a mismatch within `lerna`, causing the error.

So, what we need is a new step between points 3 and 4:
* After the lmc's compilation process is complete, its package.json is restored (possibly via `git checkout` on the exact file) to its pre-build version.
    - This will restore the original `version` value, providing the exact `version` string match that `lerna` is looking for between the two packages (`@keymanapp/lexical-model-compiler` and `@keymanapp/input-processor`).

**Comment by mcdurdin** [comment]
> * This will restore the original `version` value, providing the exact `version` string match that `lerna` is looking for between the two packages (`@keymanapp/lexical-model-compiler` and `@keymanapp/input-processor`).

The thing that bothers me here is that we are getting different version strings from different parts of the build process. Shouldn't we be getting and using the same version string everywhere? 

* @keymanapp/lexical-model-compiler: `version[-tier]` (no tier on stable)
* @keymanapp/lexical-model-types: `version[-tier]` (no tier on stable)
* @keymanapp/keyboard-processor: `version` (no tier at all)
etc.

**Comment by jahorton** [comment]
> I think that the Windows (Desktop/Developer) failures are likely due to some sort of `developer/js/package.json` shenanigans during the build process. I've occasionally seen `npm` tweak package versions during builds unexpectedly, so something might be tweaking package.json's `"version"` entry to something like `"version": "14.0.178-alpha". This could cause the "version" string to mismatch and thus cause `lerna`'s attempts to link the local development version to fail (due to version string mismatch).

I 100% agree with you.  Unfortunately, the _tooling itself_ changes the value _during_ the build in a way we can't control.

> Shouldn't we be getting and using the same version string everywhere?

Yes, absolutely.  But there's an extra detail here that's important:  "and every time".  Something in the build is invalidating this extra detail _during the build itself_.  So, we need to **revert** that "something" so that it remains the same throughout the build.

**Comment by jahorton** [comment]
That is, "sub-build 1" changes a version string before "sub-build 2" gets the chance to read in the original value.

**Comment by mcdurdin** [comment]
> Yes, absolutely. But there's an extra detail here that's important: "and every time". Something in the build is invalidating this extra detail _during the build itself_. So, we need to **revert** that "something" so that it remains the same throughout the build.

Wouldn't it be better to locate and fix the root cause of the version variance so that the version information being used is the same in all circumstances?

**Comment by jahorton** [comment]
> > It has been a while since I've built Developer locally, so maybe there's a new build prerequisite I missed?
> 
> I think you need to update your Delphi version to 10.3.3.

I'll aim to do that.  In the meantime, noting the currently-stated requirements:  https://github.com/keymanapp/keyman/blob/53a197fba592baa566b0077f3b654fb396b53ecd/windows/src/README.md#build-prerequisites

That should probably be updated; 10.2 corresponds to `DELPHI_VERSION=19.0`.

**Comment by jahorton** [comment]
Huh.  That last commit seems to have done the trick... at least, in terms of letting the build pass.  It appears that Developer runs the `lexical-model-compiler` unit tests _first_, before either of the `lexical-model-compiler`'s `npm pack` **or** the KMW build have a chance to run.  The kmlmc build needed for the unit tests is then triggering a version change.

From the TC build logs:

```
[14:39:20]
[Step 3/6] > @keymanapp/models-templates@14.0.178 build C:\BuildAgent\work\7ac43416c45637e9\keyman\common\models\templates
[14:39:20]
[Step 3/6] > tsc
[14:39:20]
[Step 3/6] 
[14:39:24]
[Step 3/6] lerna success Bootstrapped 13 packages
[14:39:25]
[Step 3/6] v14.0.178-alpha
[14:39:26]
[Step 3/6] 
[14:39:26]
[Step 3/6] > @keymanapp/lexical-model-compiler@14.0.178-alpha build C:\BuildAgent\work\7ac43416c45637e9\keyman\developer\js
[14:39:26]
[Step 3/6] > tsc
[14:39:26]
[Step 3/6] 
[14:39:32]
[Step 3/6] Typescript compilation successful.
[14:39:33]
[Step 3/6] 
[14:39:33]
[Step 3/6] > @keymanapp/lexical-model-compiler@14.0.178-alpha test C:\BuildAgent\work\7ac43416c45637e9\keyman\developer\js
[14:39:33]
[Step 3/6] > mocha
```

I've traced that section of the logs to this part of the build script:  https://github.com/keymanapp/keyman/blob/af2e4b1b6e489a14ea8baafd6bb6f2b04272d47f/developer/js/build.sh#L130-L143

So, it appears that Developer's build is setting `$publish_version`, so when running the unit-testing Makefile it's incidentally altering the package.json `version` entry that causes the break during later sub-builds.

**Comment by mcdurdin** [comment]
> That should probably be updated; 10.2 corresponds to `DELPHI_VERSION=19.0`.

See #3871.

**Comment by jahorton** [comment]
I've noted the potential followup as #3872, though I'm not sure I've got everything scoped and tagged right, given how it arises due to cross-effects among a few scopes.

**Pull request merged**